sympy/core/function.py

```diff
@@ -348,6 +348,13 @@ def _eval_subs(self, old, new):
             old == self.func and len(self.args) in new.nargs):
             return new(*[i._subs(old, new) for i in self.args])
 
+    def doit(self, **hints):
+        deep = hints.get('deep', True)
+        if deep:
+            args = [a.doit(**hints) for a in self.args]
+        else:
+            args = self.args
+        return self.func(*args, evaluate=True)
```

Comment:
For Function evaluation is already handled by `eval` so we don't need to do `func(*args)`. It would be better if all Basic classes had some kind of equivalent for `eval` (e.g. `_eval_doit`) so it isn't necessary to create a new instance when no evaluation occurs.

Reply:
I'm sorry, could you delineate some more?

Reply:
The `eval` method is a class method that takes the args and returns a new object or None. If it returns None then the object can not evaluate. Possibly that is precisely what `_eval_doit` should be.

Reply:
Then, Is this what you suggest? Seems OK.

```python
class Basic([...]):
     ...
    def doit(self, **hints):
        result = self._eval_doit(**hints)
        if result is not None:
            return result
        if self.is_Atom:
            return self
        if hints.get('deep', True):
            terms = [term.doit(**hints) if isinstance(term, Basic) 
                        else term for term in self.args]
            return self.func(*terms)
        else:
            return self

    def _eval_doit(self, **hints):
        # May be overrided by subclass
        return None
```

```python
class Add([..]):    # Example 1 for overriding _eval_doit
     ...
    def _eval_doit(self, **hints):
        if hints.get('deep', True):
            terms = [term.doit(**hints) for term in self.args]
        else:
            terms = self.args
        return self.func(*terms, evaluate=True)
```

```python
class f(Function):    # Example 2 for overriding _eval_doit
     ...
    @classmethod
    def eval(cls, x):
        return x

    def _eval_doit(self, **hints):
        if hints.get('deep', True):
            terms = [term.doit(**hints) for term in self.args]
        else:
            terms = self.args
        return self.eval(*terms)
```